File: src/leetcode/python/UniqueBinarySearchTrees_96.py
class UniqueBinarySearchTrees_96(object):
    def numTrees(self, n):
        """
        :type n: int
        :rtype: int
        """
        path = [0] * (n + 1)
        path[0] = path[1] = 1
        i = 2
        while i <= n:
            j = i
            while j >= 1:
                path[i] += path[j - 1] * path[i - j]
                j -= 1
            i += 1
        return path[n]

File: src/leetcode/java/UniqueBinarySearchTrees_96.java
package leetcode.java.medium;

public class UniqueBinarySearchTrees_96 {
  public int numTrees(int n) {
    int[] paths = new int[n + 1];
    paths[0] = paths[1] = 1;
    for (int i = 2; i <= n; i++) {
      for (int j = i; j >= 1; j--) {
        paths[i] += paths[j - 1] * paths[i - j];
      }
    }
    return paths[n];
  }
}
